packages/@uppy/aws-s3-multipart/src/index.test.js

```diff
@@ -370,4 +370,238 @@ describe('AwsS3Multipart', () => {
+        .defaultReplyHeaders({
+          'access-control-allow-headers': '*',
+          'access-control-allow-method': 'PUT',
+          'access-control-allow-origin': '*',
+          'access-control-expose-headers': 'ETag',
+        })
+        .put((uri) => uri.includes('test/upload/multitest.dat'))
+        .reply(200, '', { ETag: 'test' })
+        .persist()
+
+      const fileSize = 50 * MB
+      core.addFile({
+        source: 'jest',
+        name: 'multitest.dat',
+        type: 'application/octet-stream',
+        data: new File([new Uint8Array(fileSize)], {
+          type: 'application/octet-stream',
+        }),
+      })
+
+      await core.upload()
+      expect(createMultipartUpload).toHaveBeenCalled()
+      expect(signPart).toHaveBeenCalledTimes(10)
+      expect(completeMultipartUpload).toHaveBeenCalled()
+    })
+
+    it('preserves file metadata if upload is aborted', async () => {
+      const signPartWithAbort = jest.fn((file, partData) => {
+        expect(file.meta.createMultipartUpload).toBe(true)
+        if (partData.partNumber === 5) {
+          core.setFileMeta(file.id, {
+            ...file.meta,
+            abortingPart: partData.partNumber,
+          })
+          core.removeFile(file.id)
+        }
+
+        core.setFileMeta(file.id, {
+          ...file.meta,
+          signPart: true,
+          [`part${partData.partNumber}`]: partData.partNumber,
+        })
+
+        return {
+          url: `https://bucket.s3.us-east-2.amazonaws.com/test/upload/multitest.dat?partNumber=${partData.partNumber}&uploadId=6aeb1980f3fc7ce0b5454d25b71992&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIATEST%2F20210729%2Fus-east-2%2Fs3%2Faws4_request&X-Amz-Date=20210729T014044Z&X-Amz-Expires=600&X-Amz-SignedHeaders=host&X-Amz-Signature=test`,
+        }
+      })
+
+      core = new Core()
+        .use(AwsS3Multipart, {
+          createMultipartUpload,
+          signPart: signPartWithAbort,
+          listParts,
+          completeMultipartUpload,
+          abortMultipartUpload,
+        })
+
+      nock('https://bucket.s3.us-east-2.amazonaws.com')
+        .defaultReplyHeaders({
+          'access-control-allow-headers': '*',
+          'access-control-allow-method': 'PUT',
+          'access-control-allow-origin': '*',
+          'access-control-expose-headers': 'ETag',
+        })
+        .put((uri) => uri.includes('test/upload/multitest.dat'))
+        .reply(200, '', { ETag: 'test' })
+        .persist()
+
+      const fileSize = 50 * MB
+      core.addFile({
+        source: 'jest',
+        name: 'multitest.dat',
+        type: 'application/octet-stream',
+        data: new File([new Uint8Array(fileSize)], {
+          type: 'application/octet-stream',
+        }),
+      })
+
+      await core.upload()
+      expect(createMultipartUpload).toHaveBeenCalled()
+      expect(signPartWithAbort).toHaveBeenCalled()
+      expect(abortMultipartUpload).toHaveBeenCalled()
```

Comment:
hm, okay. that test passes for me, so maybe it's not deterministic as written? also, just to confirm, it's `abortMultipartUpload` that's not getting called?

I'll take a look at the code and see if I can figure out why that might be failing.

in the meantime, if you have any thoughts on asynchronous logic that might be causing a race condition, please let me know; if not, that's fine too. also, is `core.removeFile(file.id)` an appropriate way to trigger that abort?

Reply:
> `abortMultipartUpload` that's not getting called?

Yes correct, the failure points to that line, and commenting it does make the test pass.

> `core.removeFile(file.id)` an appropriate way to trigger that abort?

Yes I'd say so.

Reply:
@aduh95 can you share any more information about the environment the tests should run in? I'm wondering if a brief timeout or a `setTimeout(cb, 0)` to clear the event loop would address that failure, but it's hard to know without being able to reproduce it.

Reply:
I find it surprising that you are not able to reproduce it, could you share your workflow for running the tests maybe? Here's how I do it:

```sh
corepack yarn install
corepack yarn build
corepack yarn test:unit
```

> can you share any more information about the environment the tests should run in?

Since CI is seeing the same failure as I do locally, maybe comparing your environment with the CI will be easier to find what the discrepancy might be?

Reply:
okay, was able to reproduce it. not quite sure exactly why it wasn't happening before, but setting my Node version to 18, using corepack (I think not using corepack may have been the issue?), and rebasing against the upstream repo gave me the test failure.

anyway, the specific issue was that `#setS3MultipartState` was being called after the file was removed from state in `Uppy.removeFiles`. since that call threw an error, `abortFileUpload` never reached the part where it called the custom function. adding a null/undefined check for that file when refreshed from the store now allows the pass to test.  I'm somewhat surprised that this issue came up with this change, but hoping that this fix here is the appropriate way to resolve it.

I'll watch the PR for any new CI issues and address them as they come up.